Implement circuit breaker

MitulShah1 opened the pull request:
This PR contains Circuit Breaker Middleware

@aldas @jrhrmsll  @lammel  Would appreciate it if you can review this PR.

aldas commented on circuitbreaker/README.md:
```diff
@@ -0,0 +1,58 @@
+	}
+
+	e := echo.New()
+	e.Use(circuitbreaker.CircuitBreakerMiddleware(cbConfig))
```
This example would be better if middleware is added to the route.

MitulShah1 replied:
Updated

aldas commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -0,0 +1,170 @@
+	failureCount int
+	successCount int
+	state        CircuitBreakerState
+	mutex        sync.Mutex
```
[sync.RWMutex](https://pkg.go.dev/sync#RWMutex) would be probably more appropriate here

MitulShah1 replied:
Thanks, Updated

aldas commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -0,0 +1,170 @@
+	ResetTimeout time.Duration                // Interval for monitoring the circuit state
+	SuccessReset int                          // Number of successful attempts to move to closed state
+	OnOpen       func(ctx echo.Context) error // Callback for open state
+	OnHalfOpen   func(ctx echo.Context) error // Callback for half-open state
```
some of these callbacks are not used

MitulShah1 replied:
Added

aldas commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -0,0 +1,170 @@
+				cb.failureCount = 0 // Reset failure count
+			}
+			cb.mutex.Unlock()
+		case <-cb.exitChan:
```
where is this chan used so we could receive something here?

MitulShah1 commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -0,0 +1,170 @@
+func (cb *CircuitBreaker) monitorReset() {
+	for {
+		select {
+		case <-time.After(cb.resetTimeout):
```
Thanks changed to ticker

MitulShah1 commented:
@aldas I've made changes, can you please check again?

aldas commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -0,0 +1,413 @@
+func (cb *CircuitBreaker) AllowRequest() (bool, State) {
+	atomic.AddInt64(&cb.totalRequests, 1)
+
+	cb.mutex.RLock()
```
I think guarding this small block with read lock makes little difference. This whole method should be behind read lock if it is necessary that limits are honored/guarded against races during extreme peaks.

aldas commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -0,0 +1,413 @@
+func (cb *CircuitBreaker) GetStateStats() map[string]interface{} {
+	state := cb.GetState()
+
+	return map[string]interface{}{
```
these stats should have their own struct with field and their types. returning map of interfaces looks like something that belongs to javascript world.

aldas commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -0,0 +1,413 @@
+}
+
+// HealthHandler returns an Echo handler for checking circuit breaker status
+func (cb *CircuitBreaker) HealthHandler() echo.HandlerFunc {
```
I think this is little bit overthinking - if there is a method that returns stats when everyone can easily build their own handlers.  assume that there will be people that want to have their own structure for output.

aldas commented on circuitbreaker/circuit_breaker.go:
```diff
@@ -0,0 +1,413 @@
+	cb.mutex.Lock()
+	defer cb.mutex.Unlock()
+
+	if cb.state != StateOpen {
```
this could be `if cb.state == StateOpen { return}` to get rid of 1 level of indentation

MitulShah1 commented:
@aldas Thanks for detail review, I made below changes

1. Removed Unused context
2. Instead of a channel-based semaphore, Added atomic counter operations
3. Timestamp-based state transition instead of using timers.
4. Updated the locking strategy in AllowRequest()
5. Simplified resource management
6. Updated README File

Please review now.

MitulShah1 commented:
@aldas  Can you please review?
